send-transaction-service/src/send_transaction_service.rs

```diff
@@ -455,20 +270,17 @@ impl SendTransactionService {
                 }
 
                 if (!transactions.is_empty()
-                    && last_batch_sent.elapsed().as_millis() as u64 >= config.batch_send_rate_ms)
-                    || transactions.len() >= config.batch_size
+                    && last_batch_sent.elapsed().as_millis() as u64 >= batch_send_rate_ms)
+                    || transactions.len() >= batch_size
                 {
                     stats
                         .sent_transactions
                         .fetch_add(transactions.len() as u64, Ordering::Relaxed);
-                    Self::send_transactions_in_batch(
-                        &tpu_address,
-                        &transactions,
-                        leader_info_provider.lock().unwrap().get_leader_info(),
-                        &connection_cache,
-                        &config,
-                        stats,
-                    );
+                    let wire_transactions = transactions
+                        .values()
+                        .map(|transaction_info| transaction_info.wire_transaction.clone())
+                        .collect::<Vec<Vec<u8>>>();
```

Comment:
Before inside `Self::send_transactions_in_batch` we created `Vec<&[u8]>` but inside send_transaction we construct `Vec<Vec<u8>>` anyways.
Here I'm more explicit with it now and require `send_transaction_in_batch` to take `Vec<Vec<u8>>`, this should not introduce any new allocations.

Reply:
correct, but could you please add a comment that explains this, something like

```
// We must clone the tx batch before sending because we're going to retry the batch until the txs
// are confirmed. See retry_thread() and process_transactions().
```

because I just wasted 5 minutes trying to understand what was happening in those
functions :D